singledecoder.go

```diff
@@ -111,21 +112,19 @@ func (c *SingleDecoder) getProgram(ctx context.Context, vt reflect.Type, wID int
 
 	var err error
 	if wType != nil {
-		if es, ok := wType.avroType.(errorSchema); ok {
+		if es, ok := wType.avroType.(errorSchema); ok && (es.invalidAfter.IsZero() || time.Now().Before(es.invalidAfter)) {
 			return nil, es.err
 		}
-	} else {
```

Comment:
You are correct. The c.programs in the beginning is the normal case where a compiled decoder for the schema is ready. 

When a complied decoder is not ready, we need to fetch it. This could fail due to network errors, busy servers etc. In these cases we need to retry the fetch later, if not the state will be bad until the program is restarted. This is where invalidAfter comes into play. The next time the schema is requested, we find wType, and this will be an error with invalidAfter, so we can decide if we are retry to retry or not. Some errors are permanent, for example if we have a response from the registry, but we cannot compile a decoder for it. In this case, invalidAfter will be IsZero(), and we will never retry. To fix this state, a new schema must be uploaded, or the program restarted. 

I have added some comments to make this more clear.